File: typescript/src/domain/money.ts
/**
 * Money value object: integer minor units plus an ISO-style currency code.
 */

import { InvalidOrder } from "./errors";

const CURRENCY_PATTERN = /^[A-Z]{3}$/;

/**
 * Shared maximum minor units (inclusive). Equal to `Number.MAX_SAFE_INTEGER`.
 */
export const MONEY_MINOR_UNITS_MAX = 9_007_199_254_740_991;

/**
 * A non-negative amount in integer minor units of a single currency.
 *
 * Currency codes are ISO-style (`^[A-Z]{3}$`), not ISO-4217 membership.
 * `ZZZ` is valid. Cross-currency operations raise `InvalidOrder`. Create
 * instances through {@link Money.create}; the constructor is private so no
 * unvalidated money can exist.
 */
export class Money {
  /**
   * Creates a validated `Money`, raising `InvalidOrder` on violation.
   *
   * @param minorUnits - non-negative integer amount in minor units.
   * @param currency - three-letter uppercase ISO-style currency code.
   * @returns the validated value object.
   * @throws InvalidOrder when the amount is negative or fractional, exceeds
   * the shared maximum, or the code is not a three-letter uppercase string.
   */
  static create(minorUnits: number, currency: string): Money {
    if (Number.isFinite(minorUnits) && minorUnits > MONEY_MINOR_UNITS_MAX) {
      throw new InvalidOrder(
        `money amount exceeds ${String(MONEY_MINOR_UNITS_MAX)}, got ${String(minorUnits)}`,
      );
    }
    if (!Number.isSafeInteger(minorUnits) || minorUnits < 0) {
      throw new InvalidOrder(
        `money amount must be a non-negative integer, got ${String(minorUnits)}`,
      );
    }
    if (!CURRENCY_PATTERN.test(currency)) {
      throw new InvalidOrder(
        `currency must be a 3-letter uppercase ISO-style code, got "${currency}"`,
      );
    }
    return new Money(minorUnits, currency);
  }

  /**
   * Amount in integer minor units; never negative.
   */
  readonly minorUnits: number;

  /**
   * Three-letter uppercase ISO-style currency code.
   */
  readonly currency: string;

  private constructor(minorUnits: number, currency: string) {
    this.minorUnits = minorUnits;
    this.currency = currency;
  }

  /**
   * Rejects cross-currency arithmetic as an invalid order state.
   */
  private requireSameCurrency(other: Money): void {
    if (this.currency !== other.currency) {
      throw new InvalidOrder(`currency mismatch: ${this.currency} vs ${other.currency}`);
    }
  }

  /**
   * Returns the sum of two amounts of the same currency.
   *
   * @param other - addend in the same currency.
   * @throws InvalidOrder on currency mismatch or overflow of the shared max.
   */
  add(other: Money): Money {
    this.requireSameCurrency(other);
    if (this.minorUnits > MONEY_MINOR_UNITS_MAX - other.minorUnits) {
      throw new InvalidOrder("money addition overflows the shared maximum");
    }
    return Money.create(this.minorUnits + other.minorUnits, this.currency);
  }

  /**
   * Returns this amount scaled by a non-negative integer multiplier.
   *
   * @param multiplier - non-negative integer scale factor.
   * @throws InvalidOrder when the multiplier is negative or fractional, or
   * the product overflows the shared maximum.
   */
  times(multiplier: number): Money {
    if (!Number.isSafeInteger(multiplier) || multiplier < 0) {
      throw new InvalidOrder(
        `multiplier must be a non-negative integer, got ${String(multiplier)}`,
      );
    }
    if (multiplier > 0 && this.minorUnits > Math.floor(MONEY_MINOR_UNITS_MAX / multiplier)) {
      throw new InvalidOrder("money scaling overflows the shared maximum");
    }
    return Money.create(this.minorUnits * multiplier, this.currency);
  }

  /**
   * Value equality: same minor units and same currency.
   *
   * @param other - the money to compare against.
   * @returns true when both fields match.
   */
  equals(other: Money): boolean {
    return this.minorUnits === other.minorUnits && this.currency === other.currency;
  }
}

File: python/src/warehouse/domain/money.py
"""Money value object: integer minor units plus an ISO-style currency code."""

from __future__ import annotations

import re
from dataclasses import dataclass

from warehouse.domain.errors import InvalidOrder

_CURRENCY_PATTERN = re.compile(r"[A-Z]{3}")
MONEY_MINOR_UNITS_MAX = 9_007_199_254_740_991


@dataclass(frozen=True, slots=True)
class Money:
    """
    A non-negative amount in integer minor units of a single currency.

    Currency codes are ISO-style (``^[A-Z]{3}$``), not ISO-4217 membership.
    ``ZZZ`` is valid. Cross-currency operations raise ``InvalidOrder``.
    """

    minor_units: int
    currency: str

    def __post_init__(self) -> None:
        """Validate the non-negative amount, shared maximum, and currency."""
        if isinstance(self.minor_units, bool) or not isinstance(self.minor_units, int):
            raise InvalidOrder(
                f"money amount must be an integer, got {self.minor_units!r}",
            )
        if self.minor_units < 0:
            raise InvalidOrder(
                f"money amount must be non-negative, got {self.minor_units}",
            )
        if self.minor_units > MONEY_MINOR_UNITS_MAX:
            raise InvalidOrder(
                f"money amount exceeds {MONEY_MINOR_UNITS_MAX}, got {self.minor_units}",
            )
        if not _CURRENCY_PATTERN.fullmatch(self.currency):
            raise InvalidOrder(
                f"currency must be a 3-letter uppercase ISO-style code, "
                f"got {self.currency!r}",
            )

    def add(self, other: Money) -> Money:
        """Return the sum of two amounts of the same currency."""
        self._require_same_currency(other)
        total = self.minor_units + other.minor_units
        if total > MONEY_MINOR_UNITS_MAX:
            raise InvalidOrder("money addition overflows the shared maximum")
        return Money(minor_units=total, currency=self.currency)

    def times(self, multiplier: int) -> Money:
        """Return this amount scaled by a non-negative integer multiplier."""
        if isinstance(multiplier, bool) or not isinstance(multiplier, int):
            raise InvalidOrder(f"multiplier must be an integer, got {multiplier!r}")
        if multiplier < 0:
            raise InvalidOrder(f"multiplier must be non-negative, got {multiplier}")
        product = self.minor_units * multiplier
        if product > MONEY_MINOR_UNITS_MAX:
            raise InvalidOrder("money scaling overflows the shared maximum")
        return Money(minor_units=product, currency=self.currency)

    def _require_same_currency(self, other: Money) -> None:
        """Reject cross-currency arithmetic as an invalid order state."""
        if self.currency != other.currency:
            raise InvalidOrder(
                f"currency mismatch: {self.currency} vs {other.currency}",
            )
